Python/Product/Debugger/Debugger/PythonProcess.cs

```diff
@@ -1131,201 +907,176 @@ public PythonProcess(PythonLanguageVersion languageVersion, string exe, string a
-                    _stream.WriteInt32(unboundBreakpoint.Id);
-                    if (unboundBreakpoint.IsDjangoBreakpoint) {
-                        _stream.WriteString(unboundBreakpoint.Filename);
-                    }
-                }
+        internal async Task DisableBreakpointAsync(PythonBreakpoint unboundBreakpoint, CancellationToken ct) {
+            if (HasExited) {
+                return;
             }
-        }
 
-        internal void ConnectRepl(int portNum) {
-            DebugWriteCommand("Connect Repl");
-            lock (_streamLock) {
-                _stream.Write(ConnectReplCommandBytes);
-                _stream.WriteInt32(portNum);
-            }
+            await SendDebugRequestAsync(new LDP.RemoveBreakpointRequest() {
+                language = unboundBreakpoint.IsDjangoBreakpoint ? LDP.LanguageKind.Django : LDP.LanguageKind.Python,
+                breakpointId = unboundBreakpoint.Id,
+                breakpointFileName = unboundBreakpoint.Filename,
+                breakpointLineNo = unboundBreakpoint.LineNo,
+            }, ct);
         }
 
-        internal void DisconnectRepl() {
-            DebugWriteCommand("Disconnect Repl");
-            lock (_streamLock) {
-                if (_stream == null) {
-                    return;
-                }
+        internal async Task ConnectReplAsync(int portNum, CancellationToken ct = default(CancellationToken)) {
+            await SendDebugRequestAsync(new LDP.ConnectReplRequest() {
+                portNum = portNum,
+            }, ct);
+        }
 
-                try {
-                    _stream.Write(DisconnectReplCommandBytes);
-                } catch (IOException) {
-                } catch (ObjectDisposedException) {
-                    // If the process has terminated, we expect an exception
-                }
+        internal async Task DisconnectReplAsync(CancellationToken ct = default(CancellationToken)) {
+            try {
+                await SendDebugRequestAsync(new LDP.DisconnectReplRequest() {
+                }, ct);
+            } catch (IOException) {
+            } catch (ObjectDisposedException) {
+                // If the process has terminated, we expect an exception
             }
         }
 
-        internal bool SetLineNumber(PythonStackFrame pythonStackFrame, int lineNo) {
+        internal async Task<bool> SetLineNumberAsync(PythonStackFrame pythonStackFrame, int lineNo, CancellationToken ct) {
             if (_stoppedForException) {
                 return false;
             }
 
-            DebugWriteCommand("Set Line Number");
-            lock (_streamLock) {
-                _setLineResult = false;
-                _stream.Write(SetLineNumberCommand);
-                _stream.WriteInt64(pythonStackFrame.Thread.Id);
-                _stream.WriteInt32(pythonStackFrame.FrameId);
-                _stream.WriteInt32(lineNo);
-            }
+            var timeoutSource = new CancellationTokenSource(2000);
```

Comment:
>CancellationTokenSource(2000) [](start = 36, length = 29)

Timeouts via CTSs make me nervous - can you run a test pass with timeouts set to 0 to make sure that we don't have unhandled exceptions leaking out? (You should also use the CancellationTokens static class here, which will prevent cancellations while the debugger is attached so we can step through this code. It should also make it easier for you to return an "already cancelled" token for testing purposes.)